samples/nrf9160/serial_lte_modem/src/slm_at_tcpip.c

```diff
@@ -0,0 +1,716 @@
+	}
+	if (ret) {
+		LOG_ERR("Parse failed: %d", ret);
+		return ret;
+	}
+
+	ret = connect(client.sock, (struct sockaddr *)&remote,
+		 sizeof(struct sockaddr_in));
+	if (ret < 0) {
+		LOG_ERR("connect() failed: %d", -errno);
+		do_socket_close(-errno);
+		return -errno;
+	}
+
+	client.connected = true;
+	client.callback("#XTCPCONN: 1\r\n");
+	return 0;
+}
+
+static int do_tcp_send(const char *data)
+{
+	u32_t offset = 0;
+	u32_t datalen = strlen(data);
+	char buf[16];
+	int ret;
+
+	while (offset < datalen) {
+		ret = send(client.sock, data + offset,
+			   datalen - offset, 0);
+		if (ret < 0) {
+			do_socket_close(-errno);
+			LOG_WRN("send() failed: %d", -errno);
+			break;
+		}
+
+		offset += ret;
+	}
+
+	sprintf(buf, "#XTCPSEND: %d\r\n", offset);
+	client.callback(buf);
+	LOG_DBG("TCP sent");
+	return 0;
+}
+
+static int do_tcp_receive(u16_t length, u16_t time)
+{
+	int ret;
+	char data[NET_IPV4_MTU];
+	char buf[16];
+	struct timeval tmo = {
+		.tv_sec = time
+	};
+
+	ret = setsockopt(client.sock, SOL_SOCKET, SO_RCVTIMEO,
+			&tmo, sizeof(struct timeval));
+	if (ret < 0) {
+		do_socket_close(-errno);
+		LOG_ERR("setsockopt() error: %d", -errno);
+		return ret;
+	}
+
+	if (length > NET_IPV4_MTU) {
+		ret = recv(client.sock, data, NET_IPV4_MTU, 0);
+	} else {
+		ret = recv(client.sock, data, length, 0);
+	}
+	if (ret < 0) {
+		LOG_WRN("recv() error: %d", -errno);
+		do_socket_close(-errno);
+		ret = -errno;
+	} else if (ret == 0) {
+		/*
+		 * When a stream socket peer has performed an orderly shutdown,
+		 * the return value will be 0 (the traditional "end-of-file")
+		 * The value 0 may also be returned if the requested number of
+		 * bytes to receive from a stream socket was 0
+		 * In both cases, treat as normal shutdown by remote
+		 */
+		LOG_WRN("recv() return 0");
+		do_socket_close(0);
+	} else {
+		data[ret] = 0x00;
+		client.callback("#XTCPRECV: ");
+		client.callback(data);
+		client.callback("\r\n");
+		sprintf(buf, "#XTCPRECV: %d\r\n", ret);
+		client.callback(buf);
+		ret = 0;
+	}
+
+	LOG_DBG("TCP received");
+	return ret;
+}
+
+static int do_udp_init(const char *url, u16_t port)
+{
+	int ret;
+
+	if (check_for_ipv4(url, strlen(url))) {
+		ret = parse_host_by_ipv4(url, port);
+	} else {
+		ret = parse_host_by_name(url, port, SOCK_DGRAM);
+	}
+	if (ret) {
+		LOG_ERR("Parse failed: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("UDP initialized");
+	return 0;
+}
+
+static int do_udp_sendto(const char *url, u16_t port, const char *data)
+{
+	u32_t offset = 0;
+	u32_t datalen = strlen(data);
+	int ret;
+	char buf[16];
+
+	ret = do_udp_init(url, port);
+	CHECK_ERROR(ret);
+
+	while (offset < datalen) {
+		ret = sendto(client.sock, data + offset,
+			   datalen - offset, 0,
+			   (struct sockaddr *)&remote,
+			   sizeof(struct sockaddr_in));
+		if (ret <= 0) {
+			LOG_ERR("sendto() failed: %d", -errno);
+			do_socket_close(-errno);
+			return -errno;
+		}
+
+		offset += ret;
+	}
+
+	sprintf(buf, "#XUDPSENDTO: %d\r\n", offset);
```

Comment:
Same here, this string will be longer than 16 bytes.

Reply:
Like the gps part, use a global 64-byte buffer now.